### PR TITLE
Adding documentation for dealing with DLQ messages

### PR DESCRIPTION
**Pull request by prathyushpv**
**What changed?**
Adding documentation on how to deal with messages that are added to DLQs.

**Why?**
This will help operators to find messages in DLQ, the reason why it got enqueued into DLQ etc.
It also has useful tdbg commands for deleting or re-enqueuing the messages from DLQ

**How did you test it?**

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: develop/docs/dlq.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+When a task in the Temporal server encounters a terminal error, it may be enqueued in the Dead Letter Queue (DLQ). 
+To effectively manage these DLQ messages, follow these steps:
+
+1. Search `Task enqueued to DLQ` in logs and see task details.
+2. From the log trace, you can identify the offending namespace, task type, source cluster, and target cluster of the task.
+3. From the task type, get the corresponding dlq type from this map:
+    1. transfer: 1
+    2. timer: 2
+    3. replication: 3
+    4. visibility 4
```
Comment (MichaelSnowden) [inline develop/docs/dlq.md:11]: nit: Missing a semi-colon

File: develop/docs/dlq.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+4. Search `Marking task as terminally failed, will send to DLQ` in logs and find the terminal error that caused the task to be enqueued to DLQ.
```
Comment (MichaelSnowden) [inline develop/docs/dlq.md:12]: nit: to the* DLQ

File: develop/docs/dlq.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# Managing DLQ Using tdbg
+When a task in the Temporal server encounters a terminal error, it may be enqueued in the Dead Letter Queue (DLQ). 
+To effectively manage these DLQ messages, follow these steps:
+
+1. Search `Task enqueued to DLQ` in logs and see task details.
```
Comment (MichaelSnowden) [inline develop/docs/dlq.md:5]: Also mention that they will know something has been DLQ'd by the metric being incremented and then point to the metric

File: develop/docs/dlq.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+2. From the log trace, you can identify the offending namespace, task type, source cluster, and target cluster of the task.
+3. From the task type, get the corresponding dlq type from this map:
+    1. transfer: 1
+    2. timer: 2
+    3. replication: 3
+    4. visibility 4
+4. Search `Marking task as terminally failed, will send to DLQ` in logs and find the terminal error that caused the task to be enqueued to DLQ.
+5. You can list the DLQ messages using
+   `tdbg dlq --dlq-version v2 read --dlq-type {type}`. Substitute `{type}` with the integer value from step 3.
```
Comment (MichaelSnowden) [inline develop/docs/dlq.md:14]: How do they know what the message ID will be from the logs?

File: develop/docs/dlq.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# Managing DLQ Using tdbg
+When a task in the Temporal server encounters a terminal error, it may be enqueued in the Dead Letter Queue (DLQ). 
+To effectively manage these DLQ messages, follow these steps:
+
+1. Search `Task enqueued to DLQ` in logs and see task details.
+2. From the log trace, you can identify the offending namespace, task type, source cluster, and target cluster of the task.
```
Comment (MichaelSnowden) [inline develop/docs/dlq.md:6]: Mention that, if they can't find the logs, they can also just list the DLQs to see which ones are non-empty.

File: develop/docs/dlq.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+3. From the task type, get the corresponding dlq type from this map:
+    1. transfer: 1
+    2. timer: 2
+    3. replication: 3
+    4. visibility 4
+4. Search `Marking task as terminally failed, will send to DLQ` in logs and find the terminal error that caused the task to be enqueued to DLQ.
+5. You can list the DLQ messages using
+   `tdbg dlq --dlq-version v2 read --dlq-type {type}`. Substitute `{type}` with the integer value from step 3.
+
+Now these DLQ tasks can either be purged(remove from DLQ), or merged(Re-enqueue to the original queue and try to execute again).
```
Comment (MichaelSnowden) [inline develop/docs/dlq.md:16]: Mention how they can describe or cancel these jobs after they've started them

File: develop/docs/dlq.md (hunk 4 above)
Comment (prathyushpv) [reply develop/docs/dlq.md]: Added message id in DLQWriter log and mentioned that in doc.

File: develop/docs/dlq.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
+# Managing DLQ Using tdbg
+When a task in the Temporal server encounters a terminal error, it may be enqueued in the Dead Letter Queue (DLQ).
```
Comment (MichaelSnowden) [inline develop/docs/dlq.md:2]: Define terminal error here (data corruption / how it could happen).

File: develop/docs/dlq.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
+# Managing DLQ Using tdbg
+When a task in the Temporal server encounters a terminal error, it may be enqueued in the Dead Letter Queue (DLQ). 
+To effectively manage these DLQ messages, follow these steps:
+
+1. There is a metric `dlq_writes`, which will be incremented each time a message is enqueued to the DLQ.
```
Comment (MichaelSnowden) [inline develop/docs/dlq.md:5]: Add: You can use this to figure out when a task encountered a terminal error and needs manual resolution.

File: develop/docs/dlq.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+2. Search `Task enqueued to DLQ` in logs and see task details. Get the DLQ message ID from this log line.
+3. From the log trace, you can identify the offending namespace, task type, source cluster, and target cluster of the task.
```
Comment (MichaelSnowden) [inline develop/docs/dlq.md:7]: I would consolidate this with step 2. Also, I'd only include the task type, source cluster, target cluster and message ID. Also, I'd define that source cluster and target cluster are only different for replication tasks.

File: develop/docs/dlq.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+4. From the task type, get the corresponding dlq type from this map:
+    1. transfer: 1
+    2. timer: 2
+    3. replication: 3
+    4. visibility: 4
+5. You can list the DLQ messages using the command
+   `tdbg dlq --dlq-version v2 read --dlq-type {type}`. Substitute `{type}` with the integer value from step 4.
```
Comment (MichaelSnowden) [inline develop/docs/dlq.md:14]: I'd also note that you can specify `--cluster` as the source cluster for replication tasks and you can specify `--max-message-id` to be the message ID you got from the logs.

**Review by MichaelSnowden — APPROVED** [review]

**Pull request merged**